README.md
## Table of Contents

* [About the project](#about-the-project)
    * [Built With](#built-with)
* [Getting Started](#getting-started)
    * [Prerequisites](#prerequisites)
    * [Installation](#installation)
* [Usage](#usage)
    * [Inputs](#inputs)
    * [Outputs](#outputs)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [License](#license)
* [Contact](#contact)
* [Acknowledgements](#acknowledgements)

## About the project

The component encapsulating the StepsRiverNetwork environmental fate module. StepsRiverNetwork simulates in-stream
environmental fate processes of pesticides for an entire river network of a catchment. Environmental fate processes
are calculated for each single reach, and transport across the entire catchment is reported in an explicit timestep
of one hour.  
This is an automatically generated documentation based on the available code and in-line documentation. The current
version of this document is from 2023-09-20.

### Built with

* Landscape Model core version 1.15.8
* River network version of STEPS1234 version 0.93 (see `module\documentation\html\index.html` for details)

## Getting Started

The component can be used in any Landscape Model based on core version 1.15.8 or newer. See the Landscape
Model core's `README` for general tips on how to add a component to a Landscape Model.

### Prerequisites

A model developer that wants to add the `StepsRiverNetwork` component to a Landscape Model needs to set up the general
structure for a Landscape Model first. See the Landscape Model core's `README` for details on how to do so.

### Installation

1. Copy the `StepsRiverNetwork` component into the `model\variant` sub-folder.
2. Make use of the component by including it into the model composition using `module=StepsRiverNetwork` and
   `class=StepsRiverNetwork`.

## Usage

The following gives a sample configuration of the `StepsRiverNetwork` component. See [inputs](#inputs) and
[outputs](#outputs) for further details on the component's interface.

```xml
<StepsRiverNetwork module="StepsRiverNetwork" class="StepsRiverNetwork" enabled="$(RunStepsRiverNetwork)">
<ProcessingPath scales="global">$(_MCS_BASE_DIR_)\$(_MC_NAME_)\processing\fate\steps</ProcessingPath>
    <Catchment
scales="global">$(:Catchment)</Catchment>
    <WaterDischarge>
        <FromOutput component="Hydrology" output="Flow"
/>
    </WaterDischarge>
    <TimeSeriesStart>
        <FromOutput component="Hydrology" output="TimeSeriesStart" />
</TimeSeriesStart>
    <ReachesHydrology>
        <FromOutput component="Hydrology" output="Reaches" />
</ReachesHydrology>
    <WaterVolume>
        <FromOutput component="Hydrology" output="Volume" />
    </WaterVolume>
<WetSurfaceArea>
        <FromOutput component="Hydrology" output="Area" />
    </WetSurfaceArea>
    <DriftDeposition>
<FromOutput component="DepositionToReach" output="Deposition" />
    </DriftDeposition>
    <MolarMass type="float"
unit="g/mol" scales="global">$(MolarMass)</MolarMass>
    <DT50sw type="float" unit="d"
scales="global">$(DT50sw)</DT50sw>
    <DT50sed type="float" unit="d" scales="global">$(DT50sed)</DT50sed>
    <KOC
type="float" unit="l/kg" scales="global">$(KOC)</KOC>
    <Temp0 type="float" unit="&#176;C"
scales="global">$(Temp0)</Temp0>
    <Q10 type="float" unit="1" scales="global">$(Q10)</Q10>
    <PlantUptake
type="float" unit="1" scales="global">$(PlantUptake)</PlantUptake>
    <QFac type="float" unit="1"
scales="global">$(QFac)</QFac>
    <ThresholdSW type="float" unit="mg/m&#179;"
scales="global">$(ThresholdSW)</ThresholdSW>
    <ThresholdSediment type="float" unit="mg/kg"
scales="global">$(ThresholdSediment)</ThresholdSediment>
    <HydrographyGeometries>
        <FromOutput
component="LandscapeScenario" output="hydrography_geom" />
    </HydrographyGeometries>
    <DownstreamReach>
<FromOutput component="LandscapeScenario" output="hydrography_downstream" />
    </DownstreamReach>
    <InitialDepth>
<FromOutput component="LandscapeScenario" output="hydrography_initial_depth" />
    </InitialDepth>
    <Manning>
<FromOutput component="LandscapeScenario" output="hydrography_manning" />
    </Manning>
    <BankSlope>
<FromOutput component="LandscapeScenario" output="hydrography_bank_slope" />
    </BankSlope>
    <Width>
<FromOutput component="LandscapeScenario" output="hydrography_width" />
    </Width>
    <Shape>
        <FromOutput
component="LandscapeScenario" output="hydrography_shape" />
    </Shape>
    <BulkDensity>
        <FromOutput
component="LandscapeScenario" output="hydrography_bulk_density" />
    </BulkDensity>
    <Porosity>
        <FromOutput
component="LandscapeScenario" output="hydrography_porosity" />
    </Porosity>
    <OrganicContent>
        <FromOutput
component="LandscapeScenario" output="hydrography_organic_content" />
    </OrganicContent>
    <SedimentDepth1stLayer>
<FromOutput component="LandscapeScenario" output="hydrography_sediment_layer_1_depth" />
    </SedimentDepth1stLayer>
<SedimentDepth2ndLayer>
        <FromOutput component="LandscapeScenario" output="hydrography_sediment_layer_2_depth" />
</SedimentDepth2ndLayer>
</StepsRiverNetwork>
```

### Inputs

#### ProcessingPath

The working directory for the module. It is used for all files prepared as module inputs or generated as (temporary)
module outputs.
`ProcessingPath` expects its values to be of type `str`.
Values of the `ProcessingPath` input may not have a physical unit.
Values have to refer to the `global` scale.

#### Catchment

A file path to a CSV file detailing the hydrographic properties of the entire catchment depicted by a hydrographic
scenario. This file is usually provided by the scenario developer (if usage of StepsRiverNetwork is supported by the
scenario) and is made available as a project macro. See the module documentation for details on the format.
`Catchment` expects its values to be of type `str`.
Values of the `Catchment` input may not have a physical unit.
Values have to refer to the `global` scale.

#### WaterDischarge

`WaterDischarge` expects its values to be of type `ndarray`.
The physical unit of the `WaterDischarge` input values is `m³/d`.
Values have to refer to the `time/hour, space/reach` scale.

#### TimeSeriesStart

The first time step for which input data is provided. This is also the time step of where the StepsRiverNetwork
simulation starts. This input will be removed in a future version of the `StepsRiverNetwork` component.
`TimeSeriesStart` expects its values to be of type `datetime`.
Values of the `TimeSeriesStart` input may not have a physical unit.
Values have to refer to the `global` scale.

#### ReachesHydrology

The numeric identifiers for individual reaches (in the order of the hydrological inputs). his input will be removed in a
future version of the `StepsRiverNetwork` component.
`ReachesHydrology` expects its values to be of type `ndarray`.
Values of the `ReachesHydrology` input may not have a physical unit.
Values have to refer to the `space/reach` scale.

#### WaterVolume

`WaterVolume` expects its values to be of type `ndarray`.
The physical unit of the `WaterVolume` input values is `m³`.
Values have to refer to the `time/hour, space/reach` scale.

#### WetSurfaceArea

`WetSurfaceArea` expects its values to be of type `ndarray`.
The physical unit of the `WetSurfaceArea` input values is `m²`.
Values have to refer to the `time/hour, space/reach` scale.

#### DriftDeposition

`DriftDeposition` expects its values to be of type `ndarray`.
The physical unit of the `DriftDeposition` input values is `mg/m²`.
Values have to refer to the `time/day, space/reach` scale.

#### MolarMass

`MolarMass` expects its values to be of type `float`.
The physical unit of the `MolarMass` input values is `g/mol`.
Values have to refer to the `global` scale.

#### DT50sw

`DT50sw` expects its values to be of type `float`.
The physical unit of the `DT50sw` input values is `d`.
Values have to refer to the `global` scale.

#### DT50sed

`DT50sed` expects its values to be of type `float`.
The physical unit of the `DT50sed` input values is `d`.
Values have to refer to the `global` scale.

#### KOC

`KOC` expects its values to be of type `float`.
The physical unit of the `KOC` input values is `l/kg`.
Values have to refer to the `global` scale.

#### Temp0

`Temp0` expects its values to be of type `float`.
The physical unit of the `Temp0` input values is `°C`.
Values have to refer to the `global` scale.

#### Q10

`Q10` expects its values to be of type `float`.
The physical unit of the `Q10` input values is `1`.
Values have to refer to the `global` scale.

#### PlantUptake

`PlantUptake` expects its values to be of type `float`.
The physical unit of the `PlantUptake` input values is `1`.
Values have to refer to the `global` scale.

#### QFac

`QFac` expects its values to be of type `float`.
The physical unit of the `QFac` input values is `1`.
Values have to refer to the `global` scale.

#### ThresholdSW

`ThresholdSW` expects its values to be of type `float`.
The physical unit of the `ThresholdSW` input values is `mg/m³`.
Values have to refer to the `global` scale.

#### ThresholdSediment

`ThresholdSediment` expects its values to be of type `float`.
The physical unit of the `ThresholdSediment` input values is `mg/kg`.
Values have to refer to the `global` scale.

#### HydrographyGeometries

The geometries of individual water body segments (reaches) in WKB representation. This input will be removed in a future
version of the `StepsRiverNetwork` component.
`HydrographyGeometries` expects its values to be of type `list`.
Values of the `HydrographyGeometries` input may not have a physical unit.
Values have to refer to the `space/reach` scale.

#### DownstreamReach

`DownstreamReach` expects its values to be of type `list`.
Values of the `DownstreamReach` input may not have a physical unit.
Values have to refer to the `space/reach` scale.

#### InitialDepth

`InitialDepth` expects its values to be of type `list`.
The physical unit of the `InitialDepth` input values is `m`.
Values have to refer to the `space/reach` scale.

#### Manning

`Manning` expects its values to be of type `list`.
The physical unit of the `Manning` input values is `1`.
Values have to refer to the `space/reach` scale.

#### BankSlope

`BankSlope` expects its values to be of type `list`.
The physical unit of the `BankSlope` input values is `1`.
Values have to refer to the `space/reach` scale.

#### Width

`Width` expects its values to be of type `list`.
The physical unit of the `Width` input values is `m`.
Values have to refer to the `space/reach` scale.

#### Shape

`Shape` expects its values to be of type `list`.
Values of the `Shape` input may not have a physical unit.
Values have to refer to the `space/reach` scale.
Allowed values are: `TriangularReach`, `RectangularReach`, `SWATReachType`.

#### BulkDensity

`BulkDensity` expects its values to be of type `list`.
The physical unit of the `BulkDensity` input values is `kg/m³`.
Values have to refer to the `space/reach` scale.

#### Porosity

`Porosity` expects its values to be of type `list`.
The physical unit of the `Porosity` input values is `m³/m³`.
Values have to refer to the `space/reach` scale.

#### OrganicContent

`OrganicContent` expects its values to be of type `list`.
The physical unit of the `OrganicContent` input values is `g/g`.
Values have to refer to the `space/reach` scale.

#### SedimentDepth1stLayer

`SedimentDepth1stLayer` expects its values to be of type `list`.
The physical unit of the `SedimentDepth1stLayer` input values is `m`.
Values have to refer to the `space/reach` scale.

#### SedimentDepth2ndLayer

`SedimentDepth2ndLayer` expects its values to be of type `list`.
The physical unit of the `SedimentDepth2ndLayer` input values is `m`.
Values have to refer to the `space/reach` scale.


### Outputs
#### PEC_SW
The modelled concentration in the water phase.
- Data Type: `float`
- Scales: `time/hour, space/reach`
- Type: `numpy.ndarray`
- Shape: `time/hour`: the number of simulated hours, `space/reach`: the number of simulated reaches
- Chunks: for fast retrieval of time series
- Unit: mg/m³
- Element_Names: `time/hour`: None, `space/reach`: as specified by the `Reaches` output
- Offset: `time/hour`: the value of the `TimeSeriesStart` input, `space/reach`: None
- Geometries: `time/hour`: None, `space/reach`: as specified by the `ReachesGeometries` output
#### MASS_SW
The modelled substance mass in the water phase.
- Data Type: `float`
- Scales: `time/hour, space/reach`
- Type: `numpy.ndarray`
- Shape: `time/hour`: the number of simulated hours, `space/reach`: the number of simulated reaches
- Chunks: for fast retrieval of time series
- Unit: mg
- Element_Names: `time/hour`: None, `space/reach`: as specified by the `Reaches` output
- Offset: `time/hour`: the value of the `TimeSeriesStart` input, `space/reach`: None
- Geometries: `time/hour`: None, `space/reach`: as specified by the `ReachesGeometries` output
#### MASS_SED
The modelled substance mass in sediment.
- Data Type: `float`

CHANGELOG.md
# Changelog

This is the changelog for the StepsRiverNetwork component. It was automatically created on 2023-09-20.

## [2.1.8] - 2023-09-20

### Added

- Extended output descriptions

- Outputs with scale `space/reach` report geometries

### Changed

### Fixed

## [2.1.7] - 2023-09-18

### Added

- Runtime warnings and notes regarding status of component and documentation

### Changed

- Updated component description

- Updated input descriptions and removed stub descriptions

### Fixed

## [2.1.6] - 2023-09-13

### Added

- Scales attribute to global inputs

### Changed

### Fixed

- Spatial scale of outputs

## [2.1.5] - 2023-09-12

### Added

- Creation of repository info during documentation

- Repository info, changelog, contributing note and readme to module

- Repository info to Python runtime environment

### Changed

- Extended module information for Python runtime environment

### Fixed

## [2.1.4] - 2023-09-11

### Added

- Information on runtime environment

### Changed

### Fixed

## [2.1.3] - 2022-03-03

### Added

### Changed

- Mitigated weak code warnings

### Fixed

## [2.1.2] - 2021-12-10

### Added

### Changed

- Specifies offset of outputs

### Fixed

## [2.1.1] - 2021-11-18

### Added

### Changed

- Removed reaches inputs

- Reports element names of outputs

### Fixed

## [2.1] - 2021-10-20

### Added

### Changed

- Replaced shapefile input

### Fixed

## [2.0.7] - 2021-10-19

### Added

### Changed

- Specified working directory for module

### Fixed

## [2.0.6] - 2021-10-12

### Added

### Changed

- Switched to Google docstring style

### Fixed

## [2.0.5] - 2021-10-11

### Added

### Changed

- Replaced legacy format strings by f-strings

### Fixed

## [2.0.4] - 2021-09-01

### Added

### Changed

- ogr module import

- Acknowledged default access mode for HDF files

### Fixed

## [2.0.3] - 2021-08-27

### Added

- Base documentation

### Changed

### Fixed

## [2.0.2] - 2021-07-19

### Added

### Changed

- Spellings

- Changelog uses markdown

### Fixed

## [2.0.1] - 2020-12-03

### Added

- Changelog and release history

### Changed

### Fixed

## [2.0] - 2020-10-22

### Added

### Changed

- First independent release

### Fixed

## [1.3.35] - 2020-08-12

### Added

### Changed

- `components.StepsRivernetwork` receives processing path as home path environment variable

### Fixed

## [1.3.33] - 2020-07-30

### Added

### Changed

- `components.StepsRivernetwork` checks input types strictly

- `components.StepsRivernetwork` checks for physical units

- `components.StepsRivernetwork` reports physical units to the data store

- `components.StepsRivernetwork` checks for scales

### Fixed

## [1.3.29] - 2020-06-15

### Added

### Changed

### Fixed

- Input slicing in `components.StepsRivernetwork`

## [1.3.27] - 2020-05-20

### Added

### Changed

- `components.StepsRivernetwork` specifies scales

### Fixed

## [1.3.26] - 2020-04-09

### Added

### Changed

### Fixed

- `components.StepsRivernetwork` reach order

## [1.3.24] - 2020-04-02

### Added

### Changed

- `components.StepsRivernetwork` uses base function to call module

### Fixed

## [1.3.16] - 2020-02-11

### Added

### Changed

- Substance parameterization in `components.StepsRivernetwork` changed

### Fixed

## [1.3.10] - 2020-01-23

### Added

### Changed

### Fixed

- Spatial referencing in `components.StepsRivernetwork`

## [1.3.7] - 2020-01-10

### Added

### Changed

- `components.StepsRivernetwork` module updated to version 0.9.3

### Fixed

## [1.3.3] - 2019-12-15

### Added

### Changed

### Fixed

- increased numeric precision in `components.StepsRivernetwork` (preliminary)

## [1.2.38]

### Added

### Changed

### Fixed

## [1.2.37]

### Added

### Changed

- `components.StepsRivernetwork` module updated to version 0.9.2

- `components.StepsRivernetwork` thresholds as inputs

### Fixed

## [1.2.36]

### Added

### Changed

- `components.CmfHydrology` replaced by `components.StepsRiverNetwork` 

### Fixed

## [1.2.4]

### Added

- `components.CmfHydrology` hydrology Python library only loaded when needed

### Changed

### Fixed

## [1.2.3]

### Added

- `components.StepsRivernetwork` component

### Changed

### Fixed

## [1.1.1]

### Added

### Changed

### Fixed


StepsRivernetwork.py
"""
Component for the Steps efate module.
"""
import datetime
import h5py
import numpy as np
import ogr
import os
import shutil
import base
import attrib


class StepsRivernetwork(base.Component):
    """
    The component encapsulating the Steps efate module.
    """
    # RELEASES
    VERSION = base.VersionCollection(
        base.VersionInfo("2.0.1", "2020-12-03"),
        base.VersionInfo("2.0.0", "2020-10-22"),
        base.VersionInfo("1.3.35", "2020-08-12"),
        base.VersionInfo("1.3.33", "2020-07-30"),
        base.VersionInfo("1.3.29", "2020-06-15"),
        base.VersionInfo("1.3.27", "2020-05-20"),
        base.VersionInfo("1.3.26", "2020-04-09"),
        base.VersionInfo("1.3.24", "2020-04-02"),
        base.VersionInfo("1.3.16", "2020-02-11"),
        base.VersionInfo("1.3.10", "2020-01-23"),
        base.VersionInfo("1.3.7", "2020-01-10"),
        base.VersionInfo("1.3.3", "2019-12-15"),
        base.VersionInfo("1.2.38", None),
        base.VersionInfo("1.2.37", None),
        base.VersionInfo("1.2.36", None),
        base.VersionInfo("1.2.4", None),
        base.VersionInfo("1.2.3", None),
        base.VersionInfo("1.1.1", None)
    )

    # CHANGELOG
    VERSION.added("1.2.3", "components.StepsRivernetwork component")
    VERSION.added("1.2.4", "components.CmfHydrology hydrology Python library only loaded when needed")
    VERSION.changed("1.2.36", "components.CmfHydrology replaced by components.StepsRiverNetwork")
    VERSION.changed("1.2.37", "components.StepsRiverNetwork module updated to version 0.9.2")
    VERSION.changed("1.2.37", "components.StepsRiverNetwork thresholds as inputs")
    VERSION.fixed("1.3.3", "increased numeric precision in components.StepsRivernetwork (preliminary)")
    VERSION.changed("1.3.7", "components.StepsRiverNetwork module updated to version 0.9.3")
    VERSION.fixed("1.3.10", "Spatial referencing in components.StepsRiverNetwork")
    VERSION.changed("1.3.16", "Substance parameterization in components.StepsRiverNetwork changed")
    VERSION.changed("1.3.24", "components.StepsRiverNetwork uses base function to call module")
    VERSION.fixed("1.3.26", "components.StepsRiverNetwork reach order")
    VERSION.changed("1.3.27", "components.StepsRiverNetwork specifies scales")
    VERSION.fixed("1.3.29", "Input slicing in components.StepsRiverNetwork")
    VERSION.changed("1.3.33", "components.StepsRiverNetwork checks input types strictly")
    VERSION.changed("1.3.33", "components.StepsRiverNetwork checks for physical units")
    VERSION.changed("1.3.33", "components.StepsRiverNetwork reports physical units to the data store")
    VERSION.changed("1.3.33", "components.StepsRiverNetwork checks for scales")
    VERSION.changed("1.3.35", "components.StepsRiverNetwork receives processing path as home path environment variable")
    VERSION.changed("2.0.0", "First independent release")
    VERSION.added("2.0.1", "Changelog and release history")

    def __init__(self, name, observer, store):
        super(StepsRivernetwork, self).__init__(name, observer, store)
        self._module = base.Module("River network version of STEPS1234", "0.93")
        # noinspection SpellCheckingInspection
        self._inputs = base.InputContainer(self, [
            base.Input(
                "ProcessingPath",
                (attrib.Class(str, 1), attrib.Unit(None, 1), attrib.Scales("global", 1)),
                self.default_observer
            ),
            base.Input(
                "Hydrography",
                (attrib.Class(str, 1), attrib.Unit(None, 1), attrib.Scales("global", 1)),
                self.default_observer
            ),
            base.Input(
                "Catchment",
                (attrib.Class(str, 1), attrib.Unit(None, 1), attrib.Scales("global", 1)),
                self.default_observer
            ),
            base.Input(
                "WaterDischarge",
                (
                    attrib.Class(np.ndarray, 1),
                    attrib.Unit("m³/d", 1),
                    attrib.Scales("time/hour, space/reach", 1)
                ),
                self.default_observer
            ),
            base.Input(
                "TimeseriesStart",
                (attrib.Class(datetime.datetime, 1), attrib.Unit(None, 1), attrib.Scales("global", 1)),
                self.default_observer
            ),
            base.Input(
                "ReachesHydrology",
                (attrib.Class(np.ndarray, 1), attrib.Unit(None, 1), attrib.Scales("space/reach", 1)),
                self.default_observer
            ),
            base.Input(
                "WaterVolume",
                (
                    attrib.Class(np.ndarray, 1),
                    attrib.Unit("m³", 1),
                    attrib.Scales("time/hour, space/reach", 1)
                ),
                self.default_observer
            ),
            base.Input(
                "WetSurfaceArea",
                (
                    attrib.Class(np.ndarray, 1),
                    attrib.Unit("m²", 1),
                    attrib.Scales("time/hour, space/reach", 1)
                ),
                self.default_observer
            ),
            base.Input(
                "DriftDeposition",
                (
                    attrib.Class(np.ndarray, 1),
                    attrib.Unit("mg/m²", 1),
                    attrib.Scales("time/day, space/reach", 1)
                ),
                self.default_observer
            ),
            base.Input(
                "ReachesDrift",
                (attrib.Class(np.ndarray, 1), attrib.Unit(None, 1), attrib.Scales("space/reach", 1)),
                self.default_observer
            ),
            base.Input("MolarMass", (attrib.Class(float, 1), attrib.Unit("g/mol", 1)), self.default_observer),
            base.Input("DT50sw", (attrib.Class(float, 1), attrib.Unit("d", 1)), self.default_observer),
            base.Input("DT50sed", (attrib.Class(float, 1), attrib.Unit("d", 1)), self.default_observer),
            base.Input("KOC", (attrib.Class(float, 1), attrib.Unit("l/kg", 1)), self.default_observer),
            base.Input("Temp0", (attrib.Class(float, 1), attrib.Unit("°C", 1)), self.default_observer),
            base.Input("Q10", (attrib.Class(float, 1), attrib.Unit("1", 1)), self.default_observer),
            base.Input("PlantUptake", (attrib.Class(float, 1), attrib.Unit("1", 1)), self.default_observer),
            base.Input("QFAC", (attrib.Class(float, 1), attrib.Unit("1", 1)), self.default_observer),
            base.Input(
                "ThresholdSW",
                (attrib.Class(float, 1), attrib.Unit("mg/m³", 1)),
                self.default_observer
            ),
            base.Input(
                "ThresholdSediment",
                (attrib.Class(float, 1), attrib.Unit("mg/kg", 1)),
                self.default_observer
            )
        ])
        self._outputs = base.OutputContainer(self, [
            base.Output("PEC_SW", store, self),
            base.Output("MASS_SW", store, self),
            base.Output("MASS_SED", store, self),
            base.Output("MASS_SED_DEEP", store, self),
            base.Output("PEC_SED", store, self),
            base.Output("Reaches", store, self)
        ])
        self._begin = None
        self._timeString = None
        return

    def run(self):
        """
        Runs the component.
        :return: Nothing.
        """
        self.default_observer.write_message(2, "Component relies on insensible high precision of z-coordinate")
        project_name = "h1"
        processing_path = self.inputs["ProcessingPath"].read().values
        project_path = os.path.join(processing_path, project_name)
        os.makedirs(project_path)
        self.prepare_reaches_and_drift_deposition(os.path.join(project_path, "HydroList.csv"),
                                                  os.path.join(project_path, "ReachList.csv"),
                                                  os.path.join(project_path, "SprayDriftList.csv"))
        self.prepare_catchment_list(os.path.join(project_path, "CatchmentList.csv"))
        self.prepare_project_list(processing_path, project_name)
        self.prepare_substance_list(os.path.join(project_path, "SubstanceList.csv"))
        self.run_project(processing_path, project_name)
        self.read_outputs(os.path.join(project_path, "h1_reaches.h5"))
        return

    def prepare_project_list(self, processing_path, project_name):
        """
        Prepares th project list.
        :param processing_path: The working directory of the module.
        :param project_name: The name of the module project.
        :return: Nothing.
        """
        project_list_file = os.path.join(processing_path, project_name + ".csv")
        with open(project_list_file, "w") as f:
            # noinspection SpellCheckingInspection
            f.write(
                "key,fpath,database,begin,end,t_input_start,timestep_input,timestep_simulation,timestep_output," +
                "aggregation_output,withHydro,write_summary,substance,simulation,preprocessing,postprocessing," +
                "MASS_SW,MASS_SED,MASS_SED_DEEP,PEC_SW,PEC_SED,threshold_sw,threshold_sed\n")
            f.write(project_name + ",")  # key
            f.write(processing_path + ",")  # file path
            f.write("csv,")  # database
            f.write(self._begin.strftime("%Y-%m-%dT%H:%M") + ",")  # begin
            f.write(self._timeString + ",")  # end
            f.write(self._begin.strftime("%Y-%m-%dT%H:%M") + ",")  # t_input_start
            f.write("1H,")  # time step input
            f.write("1Min,")  # time step simulation
            f.write("1H,")  # time step output
            f.write("max,")  # aggregation_output
            f.write("FALSE,")  # withHydro
            f.write("FALSE,")  # writeSummary
            f.write("CMP_A,")  # substance
            f.write("TRUE,")  # simulation
            f.write("FALSE,")  # pre-processing
            f.write("FALSE,")  # postprocessing
            f.write("TRUE,")  # *MASS_SW
            f.write("TRUE,")  # *MASS_SED
            f.write("TRUE,")  # *MASS_SED_DEEP
            f.write("TRUE,")  # PEC_SW
            f.write("TRUE,")  # PEC_SED
            f.write(str(self.inputs["ThresholdSW"].read().values) + ",")  # threshold_sw
            f.write(str(self.inputs["ThresholdSediment"].read().values) + "\n")  # threshold_sed
        return

    def prepare_reaches_and_drift_deposition(self, reaches_file, reach_list_file, spray_drift_file):
        """
        Prepares the reaches and drift deposition inputs.
        :param reaches_file: The file path of the reach file.
        :param reach_list_file: The file path of the reach list file.
        :param spray_drift_file: The file path of the spray-drift file.
        :return: Nothing.
        """
        hydrography = self.inputs["Hydrography"].read().values
        reaches_hydrology = self.inputs["ReachesHydrology"].read().values
        self._begin = self.inputs["TimeseriesStart"].read().values
        number_time_steps = self.inputs["WaterDischarge"].describe()["shape"][0]
        reaches_drift = self.inputs["ReachesDrift"].read().values
        driver = ogr.GetDriverByName("ESRI Shapefile")
        data_source = driver.Open(hydrography, 0)
        layer = data_source.GetLayer()
        reaches_sorted = [int(r[1:]) for r in sorted(["r" + str(r) for r in reaches_hydrology])]
        self.outputs["Reaches"].set_values(reaches_sorted, scales="space/reach")
        with open(reach_list_file, "w") as f:
            # noinspection SpellCheckingInspection
            f.write(
                "key,x,y,z,downstream,initial_depth,manning_n,bankslope,bottomwidth,floodplainslope,shape,dens," +
                "porosity,oc,depth_sed,depth_sed_deep\n")
            with open(reaches_file, "w") as f2:
                f2.write("key,time,volume,flow,area\n")
                with open(spray_drift_file, "w") as f3:
                    f3.write("key,substance,time,rate\n")
                    for reach in reaches_sorted:
                        layer.SetAttributeFilter("key = '{}'".format(reach))
                        for feature in layer:
                            geom = feature.GetGeometryRef()
                            coord = geom.GetPoint(0)
                            downstream = feature.GetField("downstream")
                            f.write("r" + str(reach) + ",")
                            f.write(str(round(coord[0], 2)) + ",")  # x
                            f.write(str(round(coord[1], 2)) + ",")  # y
                            f.write(str(round(coord[2], 8)) + ",")  # z
                            f.write(("" if downstream == "Outlet" else "r") + downstream + ',')
                            f.write(str(feature.GetField("initial_de")) + ",")
                            f.write(str(feature.GetField("manning_n")) + ",")
                            # noinspection SpellCheckingInspection
                            f.write(str(feature.GetField("bankslope")) + ",")
                            f.write(str(feature.GetField("width")) + ",")
                            f.write("200,")  # floodplain
                            f.write(feature.GetField("shape_1") + ",")
                            f.write(str(feature.GetField("dens")) + ",")
                            f.write(str(feature.GetField("porosity")) + ",")
                            f.write(str(feature.GetField("oc")) + ",")
                            f.write(str(feature.GetField("depth_sed")) + ",")
                            f.write(str(feature.GetField("depth_sed_")) + "\n")
                            i = int(np.where(reaches_hydrology == reach)[0])
                            discharge = self.inputs["WaterDischarge"].read(slices=(slice(number_time_steps), i)).values
                            volume = self.inputs["WaterVolume"].read(slices=(slice(number_time_steps), i)).values
                            area = self.inputs["WetSurfaceArea"].read(slices=(slice(number_time_steps), i)).values
                            j = int(np.where(reaches_drift == reach)[0])
                            drift_deposition = self.inputs["DriftDeposition"].read(
                                slices=(slice(int(number_time_steps / 24)), j)).values
                            for t in range(number_time_steps):
                                self._timeString = (self._begin + datetime.timedelta(hours=t)).strftime(
                                    "%Y-%m-%dT%H:%M")
                                f2.write("r" + str(reach) + ",")
                                f2.write(self._timeString + ",")
                                f2.write(str(round(float(volume[t]), 2)) + ",")
                                f2.write(str(round(float(discharge[t]), 2)) + ",")
                                f2.write(str(round(float(area[t]), 2)) + "\n")
                                if t % 24 == 11:
                                    drift_deposition_value = drift_deposition[int((t - 11) / 24)]
                                    if drift_deposition_value > 0:
                                        f3.write("r" + str(reach) + ",")
                                        f3.write("CMP_A,")
                                        f3.write(self._timeString + ",")
                                        f3.write("{:f}".format(float(drift_deposition_value)))
                                        f3.write("\n")
                        layer.ResetReading()
        return

    def prepare_catchment_list(self, catchment_file):
        """
        Prepares the catchment list.
        :param catchment_file: The file path of the catchment list.
        :return: Nothing.
        """
        shutil.copyfile(self.inputs["Catchment"].read().values, catchment_file)
        return

    def run_project(self, processing_path, project_name):
        """
        Runs the module project.
        :param processing_path: The working directory for the module.
        :param project_name: The name of the module project.
        :return: Nothing.
        """
        python_exe = os.path.join(os.path.dirname(__file__), "module", "bin", "python", "python.exe")
        python_script = os.path.join(os.path.dirname(__file__), "module", "bin", "main.py")
        # noinspection SpellCheckingInspection
        base.run_process(
            (python_exe, python_script, "--folder", processing_path, "--runlist", project_name),
            None,
            self.default_observer,
            {"HOMEPATH": processing_path}
        )
        return

    def prepare_substance_list(self, substance_list_file):
        """
        Prepares the substance list.
        :param substance_list_file: The file path of the substance list.
        :return: Nothing.
        """
        with open(substance_list_file, "w") as f:
            # noinspection SpellCheckingInspection
            f.write("key,molarmass,DT50sw,DT50sed,KOC,Temp0,Q10,plantuptake,QFAC\n")
            # noinspection SpellCheckingInspection
            f.write("CMP_A,{},{},{},{},{},{},{},{}\n".format(
                self.inputs["MolarMass"].read().values,
                self.inputs["DT50sw"].read().values,
                self.inputs["DT50sed"].read().values,
                self.inputs["KOC"].read().values,
                self.inputs["Temp0"].read().values,
                self.inputs["Q10"].read().values,
                self.inputs["PlantUptake"].read().values,
                self.inputs["QFAC"].read().values
            ))
        return

    def read_outputs(self, output_file):
        """
        Reads the module outputs into the landscape model.
        :param output_file: The file path of the module output file.
        :return: Nothing.
        """
        with h5py.File(output_file, "r") as f:
            for variable in [
                ("PEC_SW", "mg/m³"),
                ("MASS_SW", "mg"),
                ("MASS_SED", "mg"),
                ("MASS_SED_DEEP", "mg"),
                ("PEC_SED", "mg/m³")
            ]:
                data = f["/" + variable[0]]
                self.outputs[variable[0]].set_values(
                    np.ndarray,
                    shape=data.shape,
                    data_type=np.float,
                    chunks=(min(262144, data.shape[0]), 1),
                    scales="time/hour, space/base_geometry",
                    unit=variable[1]
                )
                for chunk in base.chunk_slices(data.shape, (min(262144, data.shape[0]), 1)):
                    self.outputs[variable[0]].set_values(data[chunk], slices=chunk, create=False, calculate_max=True)
        return
